README.md
# [GoSlow](https://github.com/g00ds0n/goslow)

GoSlow is a video booth using an iPad and a GoPro camera to capture short video clips without having to press the buttons on the camera directly. It's called GoSlow because it was originally designed to capture high speed video off the GoPro to create slow motion videos.

## Installation

See [Installation Instructions](https://github.com/g00ds0n/goslow/wiki/Installation)

## Goal of this Project

Make an automated video booth that can be left unattended. The user should select a single button to get things started. This will automate the camera to capture a short clip. If possible, playback the video immediately afterwards.

## How it works

GoSlow has been used with a GoPro Hero 3 Black and an iPad 2 running GoSlow inside of the [Kiosk Pro](http://www.kioskproapp.com/itunes-apps) app. It may be possible to run GoSlow on any device (e.g. laptops, tablets, smartphones, etc) with a WiFi connection and web browser, but your mileage may vary.

This code is comprised of HTML, CSS, and JavaScript. The JavaScript sends commands to the camera using Ajax and some properly formatted URLs. GoSlow can either be put on the SD card of the camera or run locally on the device that is interfacing the camera.

> Note: GoPro doesn't actually support the webserver on their cameras, so it may or may not be working properly. Ideally you would want to put this code on the camera to avoid cross-domain issues when sending commands (via ajax) to the camera. Some browsers allow that to work while others do not.

### The Camera's Webserver

After you connect your device's WiFi to the camera, the webserver can be found at ``http://10.5.5.9:8080``. Here you can browse through the files on the SD card using any web browser. GoSlow browses through this webserver to playback video after one is recorded.

### The Camera's API

> Note: This is not supported by GoPro and could change at any firmware update.

We can do anything that the official GoPro app can do because the commands it sends to the camera are over HTTP. For instance, the following URL call will turn the camera on:

```
http://10.5.5.9:80/bacpac/PW?t=<password>&p=%01
```

* We use the same IP address as the webserver but we use port 80 instead.
* 'bacpac' can also be replaced with 'camera' for other commands.
* 'PW' is the command and can be a number of different two letter combinations, such as 'SH' or 'PV'.
* '&lt;password&gt;' is the WiFi password of the camera.
* '%01' is the code to turn the camera on. If we used '%00' instead, it would turn the camera off.

[More about Commands](https://github.com/g00ds0n/goslow/wiki/Commands)

### Running GoSlow on the SD Card

> Note: The webserver on the camera doesn't always work properly.

If you have a properly running webserver on the camera and you have GoSlow on the SD card, then you can access GoSlow by navigating to index.html (e.g. http://10.5.5.9:8080/goslow/index.html). You can use an app like [Kiosk Pro Lite](http://www.kioskproapp.com/itunes-apps) to run GoSlow when it's on the SD card. If this doesn't work you'll have to run GoSlow locally on the device itself.

> Note: Kiosk Pro Lite and Kiosk Pro can be used to prevent users from navigating away from GoSlow

### Running Locally on the Device

Cross-domain scripting is an issue, so pick a browser that allows this. I used the [Kiosk Pro](http://www.kioskproapp.com/itunes-apps) paid app on the iPad because it also allowed me to copy my files over to the app and run it locally. You can use a program like [DiskAid](http://www.digidna.net/diskaid) to transfer the files to Kiosk Pro.

> Note: Kiosk Pro allows autoplay for videos while Safari on iOS does not. Autoplay is necessary to automatically run the preview screen or any other part of this app that plays video.


CHANGELOG.md
GoSlow Changelog
================

v1.2.0
------
**IMPORTANT:** ``goslow-config.js`` has changed. Make sure to update accordingly.

* Updates the home screen to be image/video centric.
* No longer need to specify the video folder, just the DCIM directory. It will automatically find the latest folder.

v1.1.1
------
Hot fix release to rectify a problem with the 'loadedalldata' event. So this uses the 'play' event in its place.

v1.1.0
------
**IMPORTANT:** ``goslow-config.js`` has changed. Make sure to update accordingly.

* Restart the video booth automatically when an error occurs.
* The preview screen is now optional when setting ``live_timer`` to zero.
* The camera beeps only when starting and stopping the recording.
* The opening video on the home page ``clip.mp4`` is now optional by setting ``show_clip`` to false.
* The countdown screen now points to the camera. Use the ``direction`` variable to point the arrow up, down, left, or right.
* Clearer langauge on the countdown screen to convey that it isn't recording just yet.
* A new ``test_mode`` config variable that prevents GoSlow from connecting to the camera while running through the program.

v1.0.1
------
Update documentation

v1.0.0
------

Initial release of goslow.


js/goslow-functions.js
/**
 * Block the page, show a message, and reload the whole thing
 */
function error_restart(m) {
  var message = "There was a problem, please try again."
  if (m !== undefined) {
    message = m;
  }
  $('body').append('<div class="goslow-error-wrapper"></div><div class="alert alert-danger goslow-error">' + message + '</div>');
  // Block the page
  $('.goslow-error-wrapper').bind('click', function(){ return false; });

  setTimeout(function() {
    $('body').fadeOut(1000, function(){
      location.reload();
    });
  }, 4000);
}

/**
 * Get the last LRV (or MP4) file from the URL defined by goslow.videos
 */
function get_last() {
  var url = goslow.videos;
  $.ajax({
    url: url,
    async: false,
    success: function(data, status, xhr){
      url += $("tbody a.link", data).last().attr('href');
      console.log(url);
      $.ajax({
        url: url,
        async: false,
        success: function(data1, status, xhr){
          // Find the last anchor tag with the extension MP4
          var latest = $("a:contains('MP4')", data1).last().attr('href');
          // Pull the filename
          var filename = latest.substr(0, latest.indexOf('.'));
          // Find an LRV if one exists
          var lrv = $("a:contains('" + filename + ".LRV')", data1).attr('href');
          if (lrv !== undefined) {
            latest = lrv;
          }
          url += latest;
        },
        error: function(xhr, status, error){
          $('#done > h1').text("There was an error trying to find your video. Please try again.");
          var playback = videojs("gopro_playback");
          $('#done > video').remove();
          playback.dispose();
          setTimeout(function() {
            $('body').fadeOut(1000, function(){
              location.reload();
            });
          }, 9000);
        },
        dataType: 'html'
      });
    },
    error: function(xhr, status, error){
      $('#done > h1').text("There was an error trying to find your video. Please try again.");
      var playback = videojs("gopro_playback");
      $('#done > video').remove();
      playback.dispose();
      setTimeout(function() {
        $('body').fadeOut(1000, function(){
          location.reload();
        });
      }, 9000);
    },
    dataType: 'html'
  });
  return url;
}

/**
 * Commands to send to the camera
 *
 * Known commands found at
 * http://forums.openpilot.org/topic/15545-gcs-go-pro-wifi-widget/?p=168223
 * https://github.com/joshvillbrandt/GoProController/blob/master/GoProController.py
 * https://github.com/PhilMacKay/PyGoPro/blob/master/goPro.py
 */
function command(cmd, val) {
  if (goslow.test_mode) {
    return true;
  }
  var path = 'http://10.5.5.9:80/' +
    cmd + '?t=' +
    goslow.wifi;
  if (val !== undefined) {
    path += '&p=%' + val;
  }
  var code = null;
  $.ajax({
    url: path,
    async: false,
    timeout: 2000,
    success: function(data, status, xhr){
      if (status == 'success') {
        code = toHex(data);
      }
    },
    error: function(xhr, status, error){
      error_restart('There was a problem connecting to the camera. Restarting the video booth.');
    },
    dataType: 'text'
  });
  return code;
}

function beep() {
  command('camera/LL', '01');
  sleep(1000);
  command('camera/LL', '00');
}

function powerOn() {
  command('bacpac/PW', '01');
  sleep(5000);
}

function powerOff() {
   command('bacpac/PW', '00');
}

function volume(vol) {
  command('camera/BS', vol);
}
function previewOn() {
  command('camera/PV', '02');
}

function previewOff() {
  command('camera/PV', '00');
}

function fovWide() {
  command('camera/FV', '00');
}

function fovMed() {
  command('camera/FV', '01');
}

function fovNarrow() {
  command('camera/FV', '02');
}

function modeVideo() {
  command('camera/CM', '00');
}

function startCapture() {
  command('camera/SH', '01');
}

function stopCapture() {
  command('camera/SH', '00');
}

function resolution(res) {
  // 00: wvga
  // 01: 720
  // 02: 960
  // 03: 1080
  // 04: 1440
  // 05: 2.7K
  // 06: 4K
  // 07: 2.7K 17:9
  // 08: 4K 17:9

  switch(res) {
    case 'WVGA':
      command('camera/VV', '00');
      break;
    case '960':
      command('camera/VV', '02');
      break;
    case '1080':
      command('camera/VV', '03');
      break;
    default:
    case '720':
      command('camera/VV', '01');
      break;

  }
}

function frameRate(fps) {
  switch(fps) {
    case 12:  // (4K 17:9)
      command('camera/FS', '00');
      break;
    case 15:  // (4K)
      command('camera/FS', '01');
      break;
    case 24:  // (1080, 1440, 2.7K 17:9)
      command('camera/FS', '02');
      break;
    case 30:  // (1080, 1440, 2.7K)
      command('camera/FS', '04');
      break;
    case 48:  // (960, 1080, 1440)
      command('camera/FS', '05');
      break;
    case 60:  // (720, 1080)
      command('camera/FS', '07');
      break;
    case 100:  // (960)
      command('camera/FS', '08');
      break;
    case 120:  // (720)
      command('camera/FS', '09');
      break;
    case 240:  // (WVGA)
      command('camera/FS', '0a');
      break;
  }
}

var statsDef = {
  // TODO: Power
  'power': {
    'cmd': 'bacpac/se',
    'start': 18,
    'length': 2,
    'translate': {
      '00': 'off',
      '01': 'on'
    }
  },
  'volume': {
    'cmd': 'camera/se',
    'start': 19,
    'length': 1,
    'translate': {
      '0': 'Off',
      '1': '70%',
      '2': '100%'
    }
  },
  'led': {
    'cmd': 'camera/se',
    'start': 20,
    'length': 1,
    'translate': {
      '0': '0',
      '1': '2',
      '2': '4'
    }
  },
  'battery': {
    'cmd': 'camera/se',
    'start': 25,
    'length': 2,
    'translate': hexToDec
  },
  'spot_meter': {
    'cmd': 'camera/se',
    'start': 4,
    'length': 1,
    'translate': {
      '0': 'off',
      '1': 'on'
    }
  },
  'photos_remaining': {
    'cmd': 'camera/sx',
    'start': 27,
    'length': 4,
    'translate': hexToDec
  },
  // TODO: Video time remaining
  'video_time_remaining': {
    'cmd': 'camera/sx',
    'start': 27,
    'length': 4,
    'translate': hexToDec
  },
  'preview': {
    'cmd': 'camera/pv',
    'start': 0,
    'length': 2,
    'translate': {
      '00': 'off',
      '02': 'on'
    }
  },
  'recording': {
    'cmd': 'camera/sh',
    'start': 0,
    'length': 2,
    'translate': {
      '00': 'off',
      '01': 'on'
    }
  }
};

function cameraStats(stat) {
  var stats = {};
  var lookUpStats = {};
  if (stat === undefined) {
    // Load up all the stats
    lookUpStats = statsDef;
  }
  else {
    // Get one stat
    lookUpStats[stat] = statsDef[stat];
  }

  for (var s in lookUpStats) {
    var def = lookUpStats[s];
    if (typeof stats[["var"]] == "undefined") {
      stats[def.cmd] = command(def.cmd);
    }
    var raw = stats[def.cmd];
    var data = raw.substr(def.start, def.length);
    console.log(s + ": " + data);
    if (typeof def.translate === 'object' && typeof def.translate[data] != 'undefined') {
      stats[s] = def.translate[data];
    }
    else if (typeof def.translate == 'function') {
      stats[s] = def.translate(data);
    }
    else {
      stats[s] = data;
    }
  }

  return stats;
}

function toHex(str) {
  var hex = '';
  for (var i=0; i<str.length; i++) {
    hex += '' + str.charCodeAt(i).toString(16);
  }
  return hex;
}

function hexToDec(hex) {
  return parseInt(hex.toString(), 16);
}

function next_page(show) {
  $('.goslow-page').hide();
  var height = $(window).height() + "px";
  $('#' + show).css('height', height).show();
}

var auto_off = function auto_off(){
  // Unbind the click so the user can't click the screen while it's powering off
  $('#instructions').unbind('click', ready);
  command('bacpac','PW','%00');
  setTimeout(function(){
    // Rebind the home page to be click ready after 3 seconds
    $('#instructions').bind('click', ready);
  }, 3000);
};

function sleep(milliseconds) {
  var start = new Date().getTime();
  for (var i = 0; i < 1e7; i++) {
    if ((new Date().getTime() - start) > milliseconds){
      break;
    }
  }
}


js/goslow.js

$(document).ready(function() {
  $('.start-button').hide();
  // Disable rubber band effect on mac browser
  $(document).bind(
    'touchmove',
    function(e) {
      e.preventDefault();
    }
  );

  // Initially load an example video
  videojs("gopro_stream").ready(function(){
    var stream = this;
    if (goslow.show_clip) {
      stream.src({src: "clip.mp4", type: "video/mp4"})
        .on("play", function(){
          // Initialize the camera after the video loads
          preload();
        })
        .on("error", function(xhr, status, error){
          // Reload this page if there's errors with
          // loading this video
          error_restart();
        });
    }
    else {
      // Hide the spinner when using just the poster
      $('.vjs-loading-spinner').css('opacity', '0');
      // Initialize the camera after the page loads
      preload();
    }
  });

});

/**
 * Change camera's initial settings when the home page loads
 */
function preload() {
  powerOn();
  previewOff();
  stopCapture();
  volume('00');
  fovWide();
  $('.loading').hide();
  // Show user buttons after camera checks out
  $('.start-button').show().bind('click', ready);
}

/**
 * The "ready" page opens up a live stream from the camera, then
 * moves onto the "countdown" screen after some time defined by
 * "goslow.live_timer" global variable in the config file
 */
var ready = function() {
  // Show the spinner over the video
  $('.vjs-loading-spinner').css('opacity', '1');
  // Hide the button pressed and switch the other
  $(this).hide();
  $('.start-button').html($(this).html());
  $('.start-button').unbind().
    removeClass('btn-primary').removeClass('btn-warning').
    addClass('btn-danger').addClass('disabled').
    parent().removeClass('col-sm-6').addClass('col-sm-12');

  var video_type = $(this).data('video');

  // Change video resolution
  switch(video_type) {
    case "slowmo":
      resolution('720');
      frameRate(120);
      goslow.record_timer = 5;
      goslow.repeat = 0;

      $('#recording-text').html('High Speed');
      break;
    case "message":
      resolution('1080');
      frameRate(24);
      goslow.record_timer = 10;
      goslow.repeat = 1;

      $('#recording-text').html('Message');
      break;
  }

  if (goslow.live_timer > 0) {
    $('#instructions-title').fadeOut();
    $('#instructions-text').fadeOut(function(){
      var preview = '<div class="text-danger" style="font-size:2.8em">Preview';
      if (goslow.test_mode == true) {
        preview += ': Test Mode';
      }
      preview += '</div>';
      $(this).html(preview).fadeIn();
    });

    if (goslow.test_mode != true) {
      previewOn();
      videojs("gopro_stream")
        .volume(0)
        .src({src: goslow.live, type: "video/mp4"})
        .on("loadeddata", function(){
          $('.delay-message').show(800);
          setTimeout(function(){
            videojs("gopro_stream").pause();
            $('#instructions').fadeOut(function(){
              previewOff();
              videojs("gopro_stream").dispose();
              modeVideo();
              countdown();
            });
          }, goslow.live_timer * 1000);
        })
        .on("error", function(xhr, status, error){
          // Reload this page if there's errors with
          // loading the stream
          error_restart('There was a problem loading the live stream from the camera. Please try again.');
        });
    }
    else {
      setTimeout(function(){
        $('#instructions').fadeOut(function(){
          videojs("gopro_stream").dispose();
          countdown();
        });
      }, goslow.live_timer * 1000);
    }
  }
  else {
    $('#instructions').fadeOut(function(){
      videojs("gopro_stream").dispose();
      modeVideo();
      countdown();
    });
  }
};

/**
 * Starts off a countdown before recording
 */
function countdown() {
  next_page("countdown");
  // Turn up the volume on the camera so the recording beeps can be heard
  volume('02');
  // Display and animate an arrow pointing at the camera
  $("#countdown .arrow").
    html('<i class="fa fa-arrow-circle-' + goslow.direction + '"></i>').
    css('opacity', '0').animate({'opacity':1}, 500, function(){
      $("#countdown .arrow").animateRotate(360, 1500, function(){
        // Delay a second before running the countdown
        setTimeout(function(){
          // Remove the arrow to make room for the counter
          $("#countdown .arrow").animate({'opacity':0}, 900, function(){
            // Temporary fill prevents text from shifting on the screen
            $("#countdown .count").css('opacity', 0).html('J');
            $("#countdown .arrow").hide();
          });
          // Run the countdown timer
          var count = 4;
          var countdown = setInterval(function(){
            $("#countdown .count").
              css('opacity', 0).html(count).
              css('opacity', 1).animate({'opacity':0}, 900);
            count--;

            if (count == 0) {
              clearInterval(countdown);
              setTimeout(function(){
                $("#countdown").fadeOut(function(){
                  // Start recording when the count is done
                  recording();
                });
              }, 1000);
            }
          }, 1000);
        }, 1000)
      });
    });
}

function recording() {
  startCapture();
  next_page("recording");
  var count = goslow.record_timer;

  $("#recording .count").html(count).fadeOut(900);
  count--;

  var countdown = setInterval(function(){
    $("#recording .count").html(count).show().fadeOut(900);
    count--;

    if (count == -1) {
      clearInterval(countdown);
      setTimeout(function(){
        // Stop recording
        stopCapture();
        $("#recording").fadeOut(function(){
          done();
        });
      }, 1000);
    }
  }, 1000);
}

function done() {
  next_page("done");
  if (goslow.repeat > 0 && goslow.test_mode != true) {
    $('#done').append('<video id="gopro_playback" autoplay class="video-js vjs-default-skin" width="640" height="356"></video>');
    videojs("gopro_playback").ready(function(){
      var latestVideo = get_last();
      var ext = latestVideo.substr(latestVideo.lastIndexOf('.') + 1);
      var count = goslow.repeat;

      var playback = this;

      playback
        // Turn on the volume, iOS devices ignore this
        .volume(1)
        // Load the latest video
        .src({
          src: latestVideo,
          type: "video/mp4"
        })
        // Pause it to stop the autoplay
        .pause()
        // No errors occurred while loading, start off here
        .on("loadstart", function(){
          $('#playback_text').text('Getting your video...');
          // If it's an MP4, it will step through the video to show the user
          // something is happenning. This adds one more count to the number
          // of times it's repeated.
          if (ext == 'MP4') {
            //count++;
          }
        })
        // Display the percentage loaded
        .on("progress", function(){
          var loaded = this.bufferedPercent();
          // Only step through the video if it's an mp4 (Because they're huge)
          if (ext == 'MP4') {
            // Have the video step thru while it's loading, one second at a time
            var positionLoaded = Math.floor(this.duration() * loaded);
            if (positionLoaded >= this.currentTime() && loaded < 1) {
              this.currentTime(positionLoaded);
            }
          }
          $('#percent').text(Math.floor(loaded * 100) + '%');
        })
        // Video is fully transferred
        .on("loadedalldata", function(){
          setTimeout(function(){
            $('#playback_text').text("Here's your video!");
            $('#percent').text('');
          }, 500);

          // Reset the playhead
          this.currentTime(0);
          // Wait a second then start playing the video
          setTimeout(function(){
            playback.play();
            // Failsafe timeout if the video doesn't repeat correctly
            setTimeout(function(){
              $('html').fadeOut(1000, function(){
                location.reload();
              });
            }, ((goslow.repeat * goslow.record_timer) + (goslow.record_timer * 2)) * 1000);
          }, 1000);
        })
        .on("ended", function(){
          count--;
          if (count > 0) {
            if (count == 1 && goslow.playback_slowmo) {
              $('#playback_text').text("Slow motion");
              // In video.js, the video tag given a new ID
              var video = document.getElementById($('#gopro_playback video').attr('id'));
              // playbackRate isn't available on video.js so we change it a different way
              video.playbackRate = 0.5;
            }
            this.play();
          }
          else {
            setTimeout(function(){
              $('body').fadeOut(1000, function(){
                location.reload();
              });
            }, 1000);
          }
        })
        .on("error", function(xhr, status, error){
          var playback = videojs("gopro_playback");
          $('#done > video').remove();
          playback.dispose();
          goslow.playback_text = "Sorry, there was an <span class='text-danger'>error</span> trying to playback your video.";
          skipPlayback();
        });
    });
  }
  else if (goslow.repeat > 0 && goslow.test_mode == true) {
    $('#done').append('<img width="640" height="356" src="goslow.png">');
    $('#playback_text').text('Test Mode: No Playback');
    setTimeout(function(){
      $('html').fadeOut(1000, function(){
        location.reload();
      });
    }, 5000);
  }
  else {
    skipPlayback();
  }
}

function skipPlayback() {
  $('#playback_text').html(goslow.playback_text);
  $('#percent').text('');
  setTimeout(function() {
    $('body').fadeOut(1000, function(){
      location.reload();
    });
  }, 9000);
}
